message/infrastructure/sql/schema.go

```diff
@@ -0,0 +1,211 @@
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+)
+
+type SchemaAdapter interface {
+	// AckQuery returns the SQL query that will mark a message as read for a given consumer group.
+	// Subscriber will not return those messages again for this consumer group.
+	AckQuery(topic string) string
+	// AckArgs transforms the recovered message's offset and consumer group into the arguments put into AckQuery.
+	AckArgs(offset int, consumerGroup string) ([]interface{}, error)
+
+	// InsertQuery returns the SQL query that will insert the Watermill message into the SQL storage.
+	InsertQuery(topic string) string
+	// InsertArgs transforms the topic and Watermill message into the arguments put into InsertQuery.
+	InsertArgs(topic string, msg *message.Message) ([]interface{}, error)
+
+	// SelectQuery returns the SQL query that returns the next unread message for a given consumer group.
+	// Subscriber will not return those messages again for this consumer group.
+	SelectQuery(topic string) string
+	// SelectArgs transforms the topic into the argument put into SelectQuery.
+	SelectArgs(topic string, consumerGroup string) ([]interface{}, error)
+	// UnmarshalMessage transforms the Row obtained from the SQL query into a Watermill message.
+	// It also returns the offset of the last read message, for the purpose of acking.
+	UnmarshalMessage(row *sql.Row) (offset int, msg *message.Message, err error)
+
+	// EnsureTableForTopicQueries returns SQL query which will make sure (CREATE IF NOT EXISTS)
+	// that the tables exist to write messages to the given topic.
+	EnsureTableForTopicQueries(topic string) []string
+}
+
+// DefaultSchema is a default implementation of Inserter, Selecter and Acker that works with the following schema:
+//
+// `offset` bigint(20) NOT NULL AUTO_INCREMENT,
+// `uuid` binary(16) NOT NULL,
+// `created_at` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+// `payload` json DEFAULT NULL,
+// `metadata` json DEFAULT NULL,
+// `topic` varchar(255) NOT NULL,
+type DefaultSchema struct {
+	Logger watermill.LoggerAdapter
+}
+
+func (s *DefaultSchema) EnsureTableForTopicQueries(topic string) []string {
+	messagesQ := strings.Join([]string{
+		"CREATE TABLE IF NOT EXISTS `watermill_" + topic + "` (",
+		"`offset` bigint(20) NOT NULL AUTO_INCREMENT PRIMARY KEY,",
+		"`uuid` binary(16) NOT NULL,",
+		"`created_at` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,",
+		"`payload` json DEFAULT NULL,",
+		"`metadata` json DEFAULT NULL,",
+		"`topic` varchar(255) NOT NULL",
+		");",
+	}, "\n")
+
+	return []string{messagesQ}
+}
+
+func (s *DefaultSchema) InsertQuery(topic string) string {
+	if s.Logger == nil {
+		s.Logger = watermill.NopLogger{}
+	}
+
+	insertQ := strings.Join([]string{
+		`INSERT INTO`,
+		s.messagesTable(topic),
+		`(uuid, payload, metadata, topic) VALUES (?,?,?,?)`,
+	}, " ")
+
+	s.Logger.Info("Preparing query to insert messages", watermill.LogFields{
```

Comment:
couldn't all logging be done in general implementation?

Reply:
moved to publisher/subscriber 74d84d8